draft-ietf-quic-qlog-main-schema.md

```diff
@@ -651,69 +650,150 @@ JSON serialization example:
 
 ## Data {#data-field}
 
-An event's "data" field is a generic object. It contains the per-event metadata and its
-form and semantics are defined per specific sort of event. For example, data
-field value definitions for QUIC and HTTP/3 can be found in {{QLOG-QUIC}} and
-{{QLOG-H3}}.
+An event's "data" field is a generic key-value map (e.g., JSON object). It
+defines the per-event metadata that is to be logged. Its specific subfields and
+their semantics are defined per specific event type. For example, data field
+definitions for QUIC and HTTP/3 can be found in {{QLOG-QUIC}} and {{QLOG-H3}}.
 
-This field is defined here as a CDDL extension point (a "socket" or
-"plug") named `$ProtocolEventData`. Other documents MUST properly extend
-this extension point when defining new data field content options to
-enable automated validation of aggregated qlog schemas.
+In order to keep qlog fully extensible, two separate CDDL extension points
+("sockets" or "plugs") are used to fully define data fields.
 
-The only common field defined for the data field is the `trigger` field,
-which is discussed in {{trigger-field}}.
+Firstly, to allow existing data field definitions to be extended (for example by
+adding an additional field needed for a new protocol feature), a CDDL "group
+socket" is used. This takes the form of a subfield with a name of
+`* $$CATEGORY-NAME-extension`. This field acts as a placeholder that can later be
+replaced with newly defined fields by assigning them to the socket with the
+`//=` operator. Multiple extensions can be assigned to the same group socket. An
+example is shown in {{groupsocket-extension-example}}.
+
+~~~~~~~~
+; original definition in document A
+MyCategoryEventX = {
+    field_a: uint8
+
+    * $$mycategory-eventx-extension
+}
+
+; later extension of EventX in document B
+$$mycategory-eventx-extension //= (
+  ? additional_field_b: bool
+)
+
+; another extension of EventX in document C
+$$mycategory-eventx-extension //= (
+  ? additional_field_c: text
+)
+
+; if document A, B and C are then used in conjunction,
+; the combined MyCategoryEventX CDDL is equivalent to this:
+MyCategoryEventX = {
+    field_a: uint8
+
+    ? additional_field_b: bool
+    ? additional_field_c: text
+}
+~~~~~~~~
+{: #groupsocket-extension-example title="Example of using a generic CDDL group socket to extend an existing event data definition"}
```

Comment:
This example makes me realize that qlog event data field order cannot be significant. I don't think we state that anywhere already?

Reply:
Good point. Tracking that here now: https://github.com/quicwg/qlog/pull/418